### PR TITLE
sync api by pipeline

### PR DESCRIPTION
**Pull request by rokkiter**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
It realizes automatic synchronization of code and automatic synchronization of tags to api repo through gitlab's pipeline
**Which issue(s) this PR fixes**:
Fixes https://github.com/clusterpedia-io/clusterpedia/issues/47

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: hack/sync-api.sh [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+BRANCH_NAME=${raw/origin\/}
+
+
+install_filter_repo(){
+    python3 -m pip install --user git-filter-repo
+}
+
+TAG_MESSAGE=""
+# 获取 clusterpedia 的 tag message
+init_tag_message(){
```
Comment (Iceber) [inline hack/sync-api.sh:53]: This seems unnecessary to wrap as a function

File: hack/sync-api.sh [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+set -o errexit
+set -o nounset
+set -o pipefail
+
+function usage() {
+    cat <<EOF
+Build components:
+    $0 <install-filter-repo|filter-repo|sync-api>
+Build plugins:
+    $0 plugins <filter-repo>
```
Comment (Iceber) [inline hack/sync-api.sh:12]: build plugins?

What is the `$0 install-filter-repo|filter-repo`  for?

File: hack/sync-api.sh [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        usage
+    exit 1
+    fi
+        REPO_ROOT=$(pwd)
+else
+    API_ROOT="${REPO_ROOT}/staging/src/github.com/clusterpedia-io/api"
+fi
+
+if [ -z $GH_TOKEN ]; then
+    if [ -z $API_REPO ]; then
```
Comment (Iceber) [inline hack/sync-api.sh:34]: If we set `API_REPO, we don't need to set GH_TOKEN?

File: hack/sync-api.sh (hunk 2 above)
Comment (rokkiter) [reply hack/sync-api.sh:12]: I think filter-repo is a plugin, it is used in line 79 for push commit to api repo
github action does not contain filter-repo , so we need install it before use

File: hack/sync-api.sh (hunk 3 above)
Comment (rokkiter) [reply hack/sync-api.sh:34]: we must set GH_TOKEN, if not we have not auth to push to api repo

File: hack/sync-api.sh (hunk 1 above)
Comment (rokkiter) [reply hack/sync-api.sh]: thanks i have fixed it

File: hack/sync-api.sh (hunk 3 above)
Comment (Iceber) [reply hack/sync-api.sh:34]: In this judgment, if GH_TOKEN is "" but the API_REPO is not "", it seems that it does not exit, but continues to execute the subsequent logic

File: hack/sync-api.sh (hunk 2 above)
Comment (Iceber) [reply hack/sync-api.sh:12]: What will happen if I execute `./sync-api.sh plugins filter-repo`?

File: hack/sync-api.sh (hunk 3 above)
Comment (rokkiter) [reply hack/sync-api.sh:34]: yes, if GH_TOKEN is ""，need exit, it is fixed

File: hack/sync-api.sh (hunk 2 above)
Comment (rokkiter) [reply hack/sync-api.sh:12]: i misunderstand the plugins and components, it has deleted

File: hack/sync-api.sh [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      echo "push tag success~"
+    else
+      git filter-repo --subdirectory-filter staging/src/github.com/clusterpedia-io/api --force
+      git remote add origin $API_REPO
+      git push origin $BRANCH_NAME
+      echo "sync code success~"
+  fi
+}
+
+install_filter_repo
```
Comment (calvin0327) [inline hack/sync-api.sh:82]: Is it better to provide a main method？

File: hack/sync-api.sh [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+set +e; REPO_ROOT=$(git rev-parse --show-toplevel 2>/dev/null);set -e
+if [ -z $REPO_ROOT ]; then
+    if [ -z $API_ROOT ]; then
+        echo "the current directory is not in the clusterpedia api path, please set API_ROOT=<clusterpedia api path>"
+        usage
+    exit 1
+    fi
+    REPO_ROOT=$(pwd)
```
Comment (Iceber) [inline hack/sync-api.sh:23]: REPO_ROOT is not used later

File: hack/sync-api.sh [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    GH_TOKEN: github token for api repo auth.
+EOF
+}
+
+set +e; REPO_ROOT=$(git rev-parse --show-toplevel 2>/dev/null);set -e
+if [ -z $REPO_ROOT ]; then
+    if [ -z $API_ROOT ]; then
+        echo "the current directory is not in the clusterpedia api path, please set API_ROOT=<clusterpedia api path>"
+        usage
+    exit 1
```
Comment (Iceber) [inline hack/sync-api.sh:21]: align
```
        usage
        exit 1
```

File: hack/sync-api.sh [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+if [ -z $GH_TOKEN ]; then
+    echo "the github token is not in the env, please check CLUSTERPEDIA_BOT_TOKEN"
+    usage
+    exit 1
+else
+    API_REPO="https://$GH_TOKEN@github.com/clusterpedia/api.git"
+fi
+
+raw=$(git branch -r --contains $REF)
```
Comment (Iceber) [inline hack/sync-api.sh:36]: The `git <cmd>` must be in REPO, right? 
If we can't get REPO_ROOT, the `git <cmd>` will definitely fail.

File: hack/sync-api.sh [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+check_tag(){
+    git tag -d $TAGNAME
+    if [ -n "$(git ls-remote --tags origin -l $TAGNAME)" ]; then
+        echo "tag already exist, delete it before retag"
+        git push -d origin $TAGNAME
+        git tag -d $TAGNAME
+    fi
+}
+
+check_branch(){
```
Comment (Iceber) [inline hack/sync-api.sh:56]: Where is the check_branch called?

File: hack/sync-api.sh (hunk 4 above)
Comment (Iceber) [reply hack/sync-api.sh:82]: this?
```bash
main() {
    install_filter_repo

    sync_api
}
```

File: .github/workflows/sync-api.yml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  sync-api:
+    name: Sync Api
+    runs-on: ubuntu-latest
+    env:
+      WORKSPACE: ${{ github.workspace }}/src/github.com/clusterpedia-io/clusterpedia
+      GH_TOKEN: ${{ secrets.CLUSTERPEDIA_BOT_TOKEN }}
+      MESSAGE: ${{ github.event.head_commit.message }}
+      REFTYPE: ${{ github.ref_type }}
+      TAGNAME: ${{ github.ref_name }}
+      REF: ${{ github.ref }}
```
Comment (Iceber) [inline .github/workflows/sync-api.yml:16]: The environment variables related to sync-api are recommended to be placed together with sync-api

File: hack/sync-api.sh [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    BRANCH_NAME=${RAW/origin\/}
+    if [ -z $BRANCH_NAME ]; then
+        echo "can not get current branch"
+        usage
+        exit 1
+    fi
+fi
+
+if [ -z $GH_TOKEN ]; then
+    echo "the github token is not in the env, please check CLUSTERPEDIA_BOT_TOKEN"
```
Comment (Iceber) [inline hack/sync-api.sh:32]: check GH_TOKEN

File: hack/sync-api.sh [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    usage
+    exit 1
+else
```
Comment (Iceber) [inline hack/sync-api.sh:35]: No need else
```bash
if [ -z $GH_TOKEN ]; then
    echo "the github token is not in the env, please check CLUSTERPEDIA_BOT_TOKEN"
    useage
    exit 1
fi

API_REPO="https://$GH_TOKEN@github.com/clusterpedia/api.git"
```

File: hack/sync-api.sh [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    GH_TOKEN: github token for api repo auth.
+EOF
+}
+
+set +e; RAW=$(git branch -r --contains $REF 2>/dev/null);set -e
+if [ -z $RAW ]; then
+    echo "the current directory is not in the clusterpedia path"
+    usege
+    exit 1
+else
```
Comment (Iceber) [inline hack/sync-api.sh:22]: no need else
```bash
set +e; RAW=$(git branch -r --contains $REF 2>/dev/null);set -e
if [ -z $RAW ]; then
    echo "the current directory is not in the clusterpedia path"
    usege
    exit 1
fi

BRANCH_NAME=${RAW/origin\/}
if [ -z $BRANCH_NAME ]; then
    echo "can not get current branch"
    usage
    exit 1
fi

File: hack/sync-api.sh (hunk 5 above)
Comment (rokkiter) [reply hack/sync-api.sh]: it is fixed

File: hack/sync-api.sh (hunk 7 above)
Comment (rokkiter) [reply hack/sync-api.sh]: it is fixed

File: hack/sync-api.sh (hunk 8 above)
Comment (rokkiter) [reply hack/sync-api.sh]: it is unused, i had delete it

File: hack/sync-api.sh [hunk 13]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+function usage() {
+    cat <<EOF
+ENV:
+    RAW: current repo the branch which contain github action ${{ github_ref }}
```
Comment (Iceber) [inline hack/sync-api.sh:10]: RAW is not taken from an environment variable
`set +e; RAW=$(git branch -r --contains $REF 2>/dev/null);set -e`

File: hack/sync-api.sh [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    BRANCH_NAME: current branch name
+        eg. BRANCH_NAME=main
+    GH_TOKEN: github token for api repo auth.
```
Comment (Iceber) [inline hack/sync-api.sh:13]: Missing description of the TAGNAME env

File: .github/workflows/sync-api.yml (hunk 9 above)
Comment (rokkiter) [reply .github/workflows/sync-api.yml:16]: the environment variables can not get by  sync-api.sh, if placed together with run sync-api.sh

**Pull request merged**